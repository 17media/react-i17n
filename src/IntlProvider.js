import React, { PureComponent } from 'react';
import IntlMessageFormat from 'intl-messageformat';
import Aux from './Aux';
import Intl from './Intl';
import { Provider } from './IntlContext';

class IntlProvider extends PureComponent {
  static defaultProps = {
    loadingTimeout: 500, // 500 ms of blank instead of fallback message
    WrappedComponent: null, // default to not render the wrapper element
  };

  constructor(props) {
    super(props);

    this.state = {
      isTimeout: false,
    };
  }

  componentDidMount() {
    this.setTimeout();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      isTimeout: false,
    }, () => {
      this.setTimeout();
    });
  }

  setTimeout = () => {
    this.timeout = global.setTimeout(() => {
      this.setState({
        isTimeout: true,
      });
    }, this.props.loadingTimeout);
  }

  render() {
    const { locale, messages, WrappedComponent, children } = this.props;
    const intl = new Intl(locale, messages);
    const { isTimeout } = this.state;

    const context = {
      intl,
      isTimeout,
      WrappedComponent,
    };

    return (
      <Provider value={context}>
        {children}
      </Provider>
    );
  }
}

export default IntlProvider;
