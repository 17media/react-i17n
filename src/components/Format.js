import React from 'react';
import { IntlConsumer } from '../IntlContext';
import Translate from './Translate';

const Format = ({ children: value, ...props }) => (
  <IntlConsumer>
    {intl => {
      if (typeof value === 'number') {
        return intl.formatNumber(value, props);
      } else if (Date.parse(value)) {
        return intl.formatDateTime(new Date(Date.parse(value)), props);
      } else if (typeof value === 'string') {
        return <Translate {...props}>{value}</Translate>;
      }

      throw new Error(`Error parsing value ${value} in <Format> children`);
    }}
  </IntlConsumer>
);

export default Format;
