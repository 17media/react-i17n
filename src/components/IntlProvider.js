import React from 'react';
import { Provider } from '../IntlContext';
import Intl from '../Intl';

class IntlProvider extends React.PureComponent {
  static getDerivedStateFromProps(props, state) {
    if (
      state.intl.locale !== props.locale ||
      state.intl.messages !== props.messages
    ) {
      return {
        intl: new Intl(props.locale, props.messages),
      };
    }

    return null;
  }

  constructor(props) {
    super(props);

    this.state = {
      intl: new Intl(props.locale, props.messages),
    };
  }

  render() {
    const { intl } = this.state;
    const { children } = this.props;

    return <Provider value={intl}>{children}</Provider>;
  }
}

export default IntlProvider;
