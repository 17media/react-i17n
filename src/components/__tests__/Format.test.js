import React from 'react';
import { create } from 'react-test-renderer';
import IntlProvider from '../IntlProvider';
import Format from '../Format';

describe('<Format />', () => {
  const locale = 'en';
  const messages = {};

  test('format number', () => {
    const number = Math.floor(Math.random() * 1000000);

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Format>{number}</Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(
      Intl.NumberFormat(locale).format(number)
    );
  });

  test('format number with options', () => {
    const number = Math.floor(Math.random() * 1000000);

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Format style="currency" currency="TWD">
          {number}
        </Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(
      Intl.NumberFormat(locale, { style: 'currency', currency: 'TWD' }).format(
        number
      )
    );
  });

  test('format date object', () => {
    const date = new Date();

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Format>{date}</Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(
      Intl.DateTimeFormat(locale).format(date)
    );
  });

  test('format date object with options', () => {
    const date = new Date();

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Format hour="2-digit" minute="2-digit" second="2-digit">
          {date}
        </Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(
      Intl.DateTimeFormat(locale, {
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
      }).format(date)
    );
  });

  test('format date string', () => {
    const date = new Date().toString();

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Format>{date}</Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(
      Intl.DateTimeFormat(locale).format(new Date(date))
    );
  });

  test('fallback to <Translate />', () => {
    const string = 'Hello World';
    const messageID = 'id';

    const testRenderer = create(
      <IntlProvider
        locale={locale}
        messages={{ [locale]: { [messageID]: string } }}
      >
        <Format>{messageID}</Format>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(string);
  });

  test('throw error for unknown value', () => {
    // suppress console.error() call when componentDidCatch catches error
    jest.spyOn(console, 'error').mockImplementationOnce(() => {});

    const value = {};

    const expected = 'expected';

    class ErrorBoundary extends React.Component {
      state = {
        hasError: false,
      };

      componentDidCatch(error) {
        expect(error).toMatchSnapshot();

        this.setState({
          hasError: true,
        });
      }

      render() {
        return this.state.hasError ? expected : this.props.children;
      }
    }

    const testRenderer = create(
      <ErrorBoundary>
        <IntlProvider locale={locale} messages={messages}>
          <Format>{value}</Format>
        </IntlProvider>
      </ErrorBoundary>
    );

    expect(testRenderer.toJSON()).toBe(expected);

    console.error.mockRestore();
  });
});
