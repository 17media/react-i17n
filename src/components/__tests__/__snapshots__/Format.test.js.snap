// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Format /> throw error for unknown value 1`] = `[Error: Error parsing value [object Object] in <Format> children]`;
